DllExportedVariablesReader32.cpp

```diff
@@ -1,58 +1,119 @@
+	ImageDataDirectory& exportTableEntry = imageDataDirectories[0];
+
+	// No export table is available to pull from
+	if (exportTableEntry.virtualAddress == 0) {
+		return;
 	}
 
-	if (!IsDll(*stream)) { 
-		throw std::runtime_error("A file or stream was passed that was not labeled as a DLL");
+	SectionTable sectionTable = FindSectionTableContainingRva(exportTableEntry.virtualAddress);
+
+	stream.Seek(RvaToFileOffset(exportTableEntry.virtualAddress, sectionTable));
+	ExportDirectoryTable exportDirectoryTable;
+	stream.Read(exportDirectoryTable);
+
+	LoadNameTable(exportDirectoryTable.namePointerRva, sectionTable, exportDirectoryTable.numberOfNamePointers);
+	
+	std::vector<std::uint16_t> exportOrdinalTable(exportNameTable.size());
+	stream.Seek(RvaToFileOffset(exportDirectoryTable.ordinalTableRva, sectionTable));
+	stream.Read(exportOrdinalTable);
+
+	exportAddressTable.resize(exportNameTable.size());
+	stream.Seek(RvaToFileOffset(exportDirectoryTable.exportAddressTableRva, sectionTable));
+	stream.Read(exportAddressTable);
+}
+
+SectionTable DllExportedVariableReader32::FindSectionTableContainingRva(std::uint32_t rva)
+{
+	for (const SectionTable& sectionTable : sectionTables)
+	{
+		if (rva >= sectionTable.virtualAddress &&
+			rva <= sectionTable.virtualAddress + sectionTable.virtualSize)
+		{
+			return sectionTable;
+		}
 	}
 
-	stream->SeekForward(96); // Seek to Export Table entry
+	throw std::runtime_error("Provided rva value of " + std::to_string(rva) + " could not be matched to a Section Table");
+}
 
-	ImageDataDirectory exportTable;
-	stream->Read(exportTable);
+void DllExportedVariableReader32::LoadNameTable(std::uint32_t rva, const SectionTable& sectionTable, std::size_t count)
+{
+	stream.Seek(RvaToFileOffset(rva, sectionTable));
+	std::vector<uint32_t> exportNamePointerTable(count);
+	stream.Read(exportNamePointerTable);
+
+	for (std::size_t i = 0; i < exportNamePointerTable.size(); ++i)
+	{
+		stream.Seek(RvaToFileOffset(exportNamePointerTable[i], sectionTable));
+		exportNameTable.push_back(ReadNullTerminatedString());
+	}
+}
 
-	int test = exportTable.virtualAddress;
+// Reads a string until a null terminator is encountered or maxCount is reached
+std::string DllExportedVariableReader32::ReadNullTerminatedString(std::size_t maxCount)
+{
+	std::string str;
 
-	std::vector<std::string> ExportNameTable;
+	char c;
+	for (std::size_t i = 0; i < maxCount; ++i)
+	{ 
+		stream.Read(c);
+		if (c == '\0') {
+			break;
+		}
 
+		str.push_back(c);
+	}
 
+	return str;
+}
 
-	free(stream);
+// File Offset = RVA - Virtual Offset + Raw Offset.
+std::uint32_t DllExportedVariableReader32::RvaToFileOffset(std::uint32_t rva, const SectionTable& sectionTable)
+{
+	return rva - sectionTable.virtualAddress + sectionTable.pointerToRawData;
 }
 
-bool DllExportedVariableReader32::IsPortableExecutableFile(Stream::FileReader& stream)
+bool DllExportedVariableReader32::IsPortableExecutableFile()
 {
 	stream.SeekForward(0x3c);
```

Comment:
Seems like this would be more robust as an absolute seek.

Reply:
Yeah, good point.